File: schemas/podcast-platform/drizzle/ad_markers.ts
// Ad insertion markers within podcast episodes for dynamic ad placement
// See README.md for full design rationale.

import { pgTable, pgEnum, uuid, integer, timestamp, index } from "drizzle-orm/pg-core";
import { episodes } from "./episodes";

export const adMarkerTypeEnum = pgEnum("ad_marker_type", ["preroll", "midroll", "postroll"]);

export const adMarkers = pgTable(
  "ad_markers",
  {
    id: uuid("id").primaryKey().defaultRandom(),
    episodeId: uuid("episode_id").notNull().references(() => episodes.id, { onDelete: "cascade" }),
    markerType: adMarkerTypeEnum("marker_type").notNull(),
    positionMs: integer("position_ms"),
    durationMs: integer("duration_ms"),
    createdAt: timestamp("created_at", { withTimezone: true }).notNull().defaultNow(),
  },
  (table) => [
    index("idx_ad_markers_episode_id_marker_type").on(table.episodeId, table.markerType),
  ]
);

File: schemas/podcast-platform/spacetimedb/ad_markers.rs
// ad_markers: Advertisement insertion points within episodes.
// See README.md for full design rationale.

use spacetimedb::SpacetimeType;
use spacetimedb::Timestamp;

#[derive(SpacetimeType, Clone)]
pub enum MarkerType {
    Preroll, // type: String
    Midroll,
    Postroll,
}

#[spacetimedb::table(name = ad_markers, public)]
pub struct AdMarker {
    #[primary_key]
    pub id: String, // UUID
    #[index(btree)]
    pub episode_id: String, // UUID — FK → episodes.id (cascade delete)
    pub marker_type: MarkerType,
    pub position_ms: Option<i32>,
    pub duration_ms: Option<i32>,
    pub created_at: Timestamp,
}
// Composite index: (episode_id, marker_type) — not supported, enforce in reducer logic
